Added multiarch build support for teleport-operator

fheinecke opened the pull request:
This is part one of a series of PRs updating drone/dronegen to support multiarch builds.

The primary purpose of this PR is to add multiarch Teleport operator container images. This includes amd64, arm, and arm64 images under the `$MAJOR_VERSION` tag on each release.

Due to the complexity of how we handle building and publishing container images this PR includes a number of other changes:
* Removed old/unneeded Makefile targets that pertain to building images.
* Updated makefiles to use the new Dockerfile
* Refactored several Dronegen functions that are now used in multiple pipelines
* Refactored several makefile targets to support local archive building, deb building, and image building from source

Regarding multiarch container builds the following features have been added:
* Added major, minor, and canonical tags based upon release
* Separate arch-specific images/tags for each major version
* Image manifest for each major version that "points" to each arch-specific image
* Publishing to all three container repos (ECR staging, ECR production, Quay)
* Highly parallelized/efficient container image pipelines that properly build a DAG for step dependencies, executing as many as possible at once
* All images will continue to be rebuilt/update daily to ensure the latest upstream packages are installed, across the latest three major Teleport versions
* Wrote dronegen so that it should be very easy to add/remove Dockerfiles (such as the Teleport, Teleport Lab), ISAs, and container repositories

After this PR is merged I will open several much smaller PRs for Teleport and Teleport lab.

Related PR: https://github.com/gravitational/teleport/pull/15514 https://github.com/gravitational/teleport/issues/3384

Test Drone run: https://drone.platform.teleport.sh/gravitational/teleport/15845

Commit:
Added multiarch build support for teleport oss, ent, and fips

Commit:
Exported image/imageTag types

Commit:
Resigned dronegen

Commit:
Removed remainder of testing changes

Commit:
Removed changes to submodules

fheinecke committed:
Merge branch 'master' into fred/multiarch-teleport-container-images

r0mant commented on .drone.yml:
```diff
@@ -7071,8 +6785,7297 @@ steps:
+  - docker pull "drone-docker-registry:5000/teleport:$DRONE_TAG-arm64"
+  - docker tag "drone-docker-registry:5000/teleport:$DRONE_TAG-arm64" "public.ecr.aws/gravitational/teleport:$(echo
+    $DRONE_TAG | sed 's/v//' | cut -d'.' -f 1)-arm64"
+  - docker tag "drone-docker-registry:5000/teleport:$DRONE_TAG-arm64" "public.ecr.aws/gravitational/teleport:$(echo
```
Can we set "major", "major.minor" and "major.minor.patch" versions to some environment variables or write them to files to avoid having to do `sed | cut` magic everywhere?

fheinecke replied:
Fixed in [186a98c7452c185030084a93ada7d89757deafaf](https://github.com/gravitational/teleport/pull/16688/commits/186a98c7452c185030084a93ada7d89757deafaf)

r0mant commented on .drone.yml:
```diff
@@ -7071,8 +6785,7297 @@ steps:
+  commands:
+  - PACKAGE_NAME="teleport-ent"
+  - PACKAGE_VERSION="$(cat '/go/vars/full-version/v9')"
+  - apt update
```
I feel very uneasy about this script basically being embedded in drone yaml. Can it at least be factored out in a bash script or something?

fheinecke replied:
I can do this if you like however I see several downsides:
* It adds a snowflake case where the logic behind this step is not in dronegen, rather in a shell script(s)/makefile target(s)
* We lose several compile-time/dronegen-runtime checks
* It introduces another layer of complexity
* It would move a fairly large amount of Go logic into the script (see the function that generates these commands: https://github.com/gravitational/teleport/pull/16688/files#diff-4df21a264c3b32375c8d1ed01e7de5aec5d63ca23f30721113aeb45db7d42f45R455-R549)

I'm all for a making our pipelines more CI platform agnostic but I'm not sure the tradeoffs in the case are worth it. What are your thoughts?

r0mant commented on Makefile:
```diff
@@ -1123,15 +1093,20 @@ rpm:
 	cp ./build.assets/build-package.sh ./build.assets/build-common.sh $(BUILDDIR)/
 	chmod +x $(BUILDDIR)/build-package.sh
 	cd $(BUILDDIR) && ./build-package.sh -t oss -v $(VERSION) -p deb -a $(ARCH) $(RUNTIME_SECTION) $(TARBALL_PATH_SECTION)
 	if [ -f e/Makefile ]; then $(MAKE) -C e deb; fi
```
Should this line be removed from `oss-deb` target? Seems `deb` below builds enterprise deb.

fheinecke replied:
Thanks I missed that - fixed

r0mant reviewed (commented):
@fheinecke A couple of meta-comments too:

- The Drone file is MASSIVE now. Does Drone have a concept of imports/includes, or can we for example factor out all cron-related stuff into a separate Drone file? I would also try to move some things that we can into scripts/programs (I left a couple of related comments).
- Because the changeset is so big, it's pretty scary to put in production right away. Can we make sure that all new pipelines that build multiarch images have failures ignored on them so once this merges we can run for a few releases without their failures (if any) affecting anything and observe general behavior?

@logand22 Can you please give this another look? This got split from the original's @fheinecke's multi-arch PR in the attempt to land it in chunks.

logand22 reviewed (changes requested):
Thanks for including a passing build in the Comments. 

This wasn't necessarily what I had in mind for reducing the size of the PR. It seems like you have a lot of changes still that are orthogonal in goals that could be done completely separately.
1. Consolidate to a single Dockerfile
2. Add multi-arch builds
3. Refactor build pipeline to use dronegen. 

 Not including operator and teleport lab doesn't necessary reduce this PR by that much. 

Other than the small comments I left in the changed files I have questions over the following:
* Do you need to reference an _updated_ `teleport.e` ref for `teleport-ent` images to work?
* Can you include examples of `promote` event running as well as `push` events?
* Why is the `promote` event building images? Shouldn't it just be pulling and retagging existing images?
* Your updates include changes to the cronjob that automatically keep the latest 10.x.x 10.x and 10 version up-to-date daily vs. just `10` (Same for previous versions). Is this intended? 

Per @r0mant comment regarding drone file size. I believe drone supports both [templates](https://docs.drone.io/template/yaml/) and [configuration extensions](https://docs.drone.io/extensions/configuration/) that could help reduce the size. However this seems either out of scope or a prerequisite for these changes.

Commit:
Reverted dockerfile-fips change

fheinecke commented on build.assets/Dockerfile-fips:
```diff
@@ -48,7 +48,7 @@ RUN apt-get update -y --fix-missing && \
 ARG UID
 ARG GID
-RUN (groupadd ci --gid=$GID -o && useradd ci --uid=$UID --gid=$GID --create-home --shell=/bin/sh && \
+RUN (groupadd ci --gid=$GID -o && useradd ci -o --uid=$UID --gid=$GID --create-home --shell=/bin/sh && \
```
This was an unintended side effect of some local testing. Good catch - reverted

Commit:
FIxed docs wording

fheinecke commented on docs/postrelease.md:
```diff
@@ -12,8 +12,7 @@ This checklist is to be run after cutting a release.
 - [ ] Update support matrix in docs FAQ page
-- [ ] Update `CURRENT_VERSION_ROOT` and other previous versions in Drone `teleport-docker-cron` job
-  - Example: https://github.com/gravitational/teleport/pull/4602
+- [ ] Update `teleportVersions` and other previous versions in Dronegen `/dronegen/container_images.go buildContainerImagePipelines` function, and run `make dronegen`
```
Had this fixed in another commit that didn't make it over when I split this PR out. Updated

fheinecke commented:
Regarding multiple files @logand22 is right, however templates are not supported in the version of Drone we're using. See https://github.com/harness/drone/blob/master/HISTORY.md#202. It's probably sufficient for this discussion to simply say that it is not feasible to split .drone.yml up within the scope of this PR @r0mant .

That being said, I think the increasing .drone.yml size is an issue I think this would be a great discussion to have outside the scope of this PR.

Commit:
Un-exported most constants

fheinecke commented on dronegen/container_images.go:
```diff
@@ -0,0 +1,869 @@
+	StagingEcrRegion      string = "us-west-2"
+
+	LocalRegistry string = "drone-docker-registry:5000"
+)
```
Updated
I left `ProductionRegistryOrg`, `PublicEcrRegion`, and `StagingEcrRegion` exported as they may be useful in the future in other files

Commit:
Removed teleport.e makefile deb call

Commit:
Moved "sed | cut magic" to files

fheinecke commented:
@logand22 to address your comments/questions above:
> This wasn't necessarily what I had in mind for reducing the size of the PR [...]

I considered this when opening the PR however there are several reasons why I split the original PR this way:
* Splitting as suggested would require significantly more work as the original PR is essentially a complete rewrite of how we build and publish container images. If I split each feature out separately (for example, consolidating Teleport dockerfiles), then I would also need to write a non-trivial amount of logic to make the feature "fit" the existing codebase. That logic would then be completely removed in future PRs. IMO this would add a lot of work for not much gain.
* Splitting the PR by product multiplicatively reduces the amount of `.drone.yml` changes. These changes are the vast majority of the PR by line count.
* If there is an issue not caught until a "production" run (such as a promotion event) it will only affect one product, not all. This ideally will greatly reduce the blast radius of arising issues.

> Do you need to reference an updated teleport.e ref for teleport-ent images to work?

No. All logic for building teleport-ent images in the context of a drone pipeline is in dronegen. I will be opening up a separate PR that updates the `teleport.e` makefile to remove some old targets, and update existing `image`/related targets to incorporate the Dockerfile changes in this PR. In other words, the `teleport.e` PR will depend on this one but not the other way around.

> Can you include examples of promote event running as well as push events?

The linked test run (https://drone.platform.teleport.sh/gravitational/teleport/15845) was triggered on a push event to this branch. Is that what you're looking for?

Regarding testing a promote event I believe I can test this by removing the "Check if tag is prerelease" step, then promoting a dev tag. Is that sufficient? 99.99% of the logic from the test run should be identical here however testing on a `promote` trigger may be useful to catch drone vars that are not set the same when a different trigger is ran.

> Why is the promote event building images? Shouldn't it just be pulling and retagging existing images?

This was primarily done in an attempt to unify how images are built. Here's what currently happens when a release is about to be published (as I understand it):
1. A tag event triggers the `build-docker-images` pipeline, which builds container images and pushes them to staging
2. A promote event triggers `promote-docker-ecr` and `promote-docker-quay` which pull from ECR staging and push to public ECR and Quay.
3. A cron event, occuring within 24 hours of a promotion, triggers `teleport-docker-cron-ecr` and `teleport-docker-cron`. These pipelines rebuild the staging ECR, public ECR, and Quay images from scratch and push to all three registries. This push replaces the tags created in (2).

As a result, the staging images built in (1) and pulled/tagged/pushed in (2) are only available for 24 hours at most. In practice I think this is closer to less than 12 hours. By having the `promote` triggered pipeline also build images we unify how images are built between pipelines with different triggers.

That being said, I did completely forget to implement `build-docker-images` for pushing to staging on tag events for testing purposes. I'll get that added.

> Your updates include changes to the cronjob that automatically keep the latest 10.x.x 10.x and 10 version up-to-date daily vs. just 10 (Same for previous versions). Is this intended?

Yes this was 100% intended. Doing so provides several advantages:
* Our customers can "sticky" a major, minor, or canonical version across patches if they so choose
* We have a consistent image tagging scheme across products (right now teleport, teleport-lab, and teleport-operator tags don't follow the same scheme)
* There is a bug with our current pipelines where only certain tags receive base images security updates which this fixes

Does this address your comments/concerned? I'm not dead set on how the PR currently changes things - just trying to explain why I made the above decisions.

Commit:
Re-added `mkdir -pv /go/cache` to push.go

logand22 commented:
> Splitting as suggested would require significantly more work as the original PR is essentially a complete rewrite of how we build and publish container images [...] 

I haven't looked into how much work it would require to make the changes individually, so I'll default to your judgement here. I only suggested that because the "complex rewrite" broken down into smaller parts would help simplify the PR review cycle (Less changes to worry about). 

> If there is an issue not caught until a "production" run (such as a promotion event) it will only affect one product, not all. This ideally will greatly reduce the blast radius of arising issues.

Agreed. In that case why not start with Teleport lab or operator due to the smaller nature of those products? 

> No. All logic for building teleport-ent images in the context of a drone pipeline is in dronegen.

Sounds good!

> This push replaces the tags created in (2).

The tags that the promote event pushes in (2) are only for the X.X.X version images. The cron event only replaces the floating tag "X" (example: promote pushes 10.2.6 but cron only updates 10)

> As a result, the staging images built in (1) and pulled/tagged/pushed in (2) are only available for 24 hours at most.

Due to above mentioned correction. This is not the case. The images built in (1) and (2) are never overwritten. The cron-job only ever overwrites the floating tags. 

The original context around having a staging registry and a public production registry was to allow Teleport engineers the ability to test images before "promoting" them to our customers via the public production registry. This way the image the Teleport engineer tests in the staging registry are a 1 to 1 match with the one promoted as it is just the retagged staging image that is promoted. 

I'm not sure what the purpose of having both "push" and "promote" build images from scratch other than to be "fresher" than when the push event occurred. This kind of defeats the purpose of the staging registry (Teleport tag doesn't change but underlying packages do). 

> Our customers can "sticky" a major, minor, or canonical version across patches if they so choose

While I'm a fan of having a floating major version and I can see even how minor versions would be beneficial to certain customers. Teleport had previous decided on keeping patch versions immutable via https://github.com/gravitational/teleport/issues/3357#issuecomment-644452081. I also think it provides a false sense of security for our customers since we bump patch versions all the time and only the latest patch version will be up to date with this method. 

I suggest we continue to encourage our customers to use a major floating version by not offering a floating patch version and keep patch versions immutable.

Commit:
Command deterministic order fix

Commit:
Added staging-only tag pipeline

Commit:
Moved PR to teleport operator to minimize potential issue impact

r0mant commented on Makefile:
```diff
@@ -33,38 +32,65 @@ ADDFLAGS ?=
 PWD ?= `pwd`
 TELEPORT_DEBUG ?= false
 GITTAG=v$(VERSION)
-BUILDFLAGS ?= $(ADDFLAGS) -ldflags '-w -s' -trimpath
```
What's the reason for this change?

r0mant commented on Makefile:
```diff
@@ -33,38 +32,65 @@ ADDFLAGS ?=
 
 # FIPS support must be requested at build time.
-FIPS_MESSAGE := without-FIPS-support
+FIPS_MESSAGE := "without FIPS support"
```
Is this just a cosmetic change? Removing `-` in all "with-xxx-support" tags.

r0mant commented on Makefile:
```diff
@@ -270,17 +230,15 @@ all: version
 # If you are considering changing this behavior, please consult with dev team first
 .PHONY: $(BUILDDIR)/tctl
 $(BUILDDIR)/tctl:
-	GOOS=$(OS) GOARCH=$(ARCH) $(CGOFLAG) go build -tags "$(PAM_TAG) $(FIPS_TAG) $(LIBPCSCLITE_BUILD_TAG)" -o $(BUILDDIR)/tctl $(BUILDFLAGS) ./tool/tctl
```
Why did you remove `$(LIBPCSCLITE_BUILD_TAG)`?

r0mant commented on Makefile:
```diff
@@ -805,14 +776,6 @@ fix-license: $(ADDLICENSE)
-# Used prior to a release by bumping VERSION in this Makefile and then
-# running "make update-version".
-.PHONY: update-version
-update-version: version test-helm-update-snapshots
```
Why did this target get removed?

r0mant reviewed (commented):
@fheinecke It seems the Makefile here now has a bunch of seemingly unrelated changes which I don't remember seeing before.

Commit:
Updated promote to pull and push without build

Commit:
Made cron triggers not affect canonical tags

Commit:
Added check for pre-existing tags on immutable CRs

Commit:
Added immutability check to manifests

Commit:
Updated staging ecr to only apply $TIMESTAMP tag on cron triggers

Commit:
Updated triggerinfo struct to use a triggerflag struct

Commit:
Fixed makefile after git mistake

Commit:
Makefile fix

Commit:
Merge branch 'master' into fred/multiarch-teleport-container-images

fheinecke commented:
> @fheinecke It seems the Makefile here now has a bunch of seemingly unrelated changes which I don't remember seeing before.

@r0mant this was due to a `git revert` screwup on my part. All changes should be reverted.

Commit:
Merge branch 'master' into fred/multiarch-teleport-container-images

fheinecke commented:
@logand22 To address your comment:

> In that case why not start with Teleport lab or operator due to the smaller nature of those products?

100% agreed. Updated PR to be for `teleport-operator` instead of `teleport`.

> [Tagging discussion...]

Understand what you're saying. I've implemented the following to address what you brought up:
* `tag` events will only push to ECR staging
* `promote` events will pull from ECR staging then push to ECR production and Quay
* `cron` events will no longer update canonical semver tags
* Only `cron` events will append `$TIMESTAMP` to ECR - staging tags
* The pipelines will now check if a given tag exists before trying to push for ECR staging (all tags) and canonical tags (all registries). If the tag exists, a message is logged and the manifest/images are not pushed without erroring.

fheinecke commented:
Note I'll be pushing some commits to test the pipelines in Drone. For review purposes [this commit backwards](https://github.com/gravitational/teleport/pull/16688/commits/c5418b889e88b7f15ab7f1e9e18abb4d1646d5a4) are what will be merged after reverting the in-place testing changes.

logand22 commented on dronegen/container_images.go:
```diff
@@ -0,0 +1,942 @@
+func GetProductionContainerRepos() []*ContainerRepo {
+	return []*ContainerRepo{
+		NewQuayContainerRepo("PRODUCTION_QUAYIO_DOCKER_USERNAME", "PRODUCTION_QUAYIO_DOCKER_PASSWORD"),
+		NewEcrContainerRepo("PRODUCTION_TELEPORT_DRONE_USER_ECR_KEY", "PRODUCTION_TELEPORT_DRONE_USER_ECR_SECRET", ProductionRegistry, true, false, false),
```
ECR Production is not immutable. (Immutability on ECR - Public isn't available currently)

fheinecke replied:
Right - the `isImmutable` parameter is set to false

fheinecke replied:
@logand22 I think I see where the confusion is here. Which of the following is the expected behavior?
1. When pushing to production registries (i.e. `public.ecr.aws`) the push step should fail if an immutable tag (currently only canonical) exists.
2. When pushing to production registries (i.e. `public.ecr.aws`) the push step should log and continue if an immutable tag (currently only canonical) exists.
3. When pushing to production registries (i.e. `public.ecr.aws`) the push step should push regardless of if an immutable tag (currently only canonical) exists.

Currently I have (2) configured which is why you see lines in `.drone.yml` [like this](https://github.com/gravitational/teleport/pull/16688/files#diff-b54b39f1afced2465e1f3641db9d5bbf4f3a7fcf890996dfedd3c197bcb7f8c7R7825-R7851). I can implement any of these behaviors - just need to know which one is correct.

logand22 replied:
Public.ecr.aws does not support tag immutability currently, therefore there is no error when attempting to push an image tag that already exists. If you want to not overwrite a canonical image when it already exists with a sort of client-side immutability then you can although I don't have that existing in the pipeline currently. 

So both 2. and 3. could work however I don't think you should fail if a tag already exists so not 1.

logand22 reviewed (approved):
Looks pretty good to me! I think reducing it to just `teleport-operator` made things a lot easier to follow currently. I do have my opinions on whether abstracting things in `dronegen` actually improve readability or not but that won't matter when we switch to Github Actions and separate actions into separate files. 

Other than some comments I left and barring tests for `tag` (I'd like to see you test your manifest inspection lines), `promote`, and `cron` it looks good to me.

r0mant commented on dronegen/container_images.go:
```diff
@@ -0,0 +1,942 @@
+	majorVersionVarPath := path.Join(majorVersionVarDirectory, majorVersion)
+	return step{
+		Name:  fmt.Sprintf("Find the latest available semver for %s", majorVersion),
+		Image: "golang:1.18",
```
We are now using Go 1.19 in master and v11. Is there a constant defined somewhere with Go version we can use instead of hardcoding it inline here?

r0mant commented on dronegen/container_images.go:
```diff
@@ -0,0 +1,942 @@
+	// *************************************************************
+	// ****** These need to be updated on each major release. ******
+	// *************************************************************
+	latestMajorVersions := []string{"v10", "v9", "v8"}
+	branchMajorVersion := "v10"
```
Do these need to be updated now that we've cut `branch/v11`?

fheinecke replied:
Yes it does - fixed pending push

r0mant commented on dronegen/container_images.go:
```diff
@@ -0,0 +1,942 @@
+			},
+		},
+		Commands: []string{
+			"docker run --privileged --rm tonistiigi/binfmt --install all",
```
What is this `tonistiigi/binfmt` image?

fheinecke replied:
That container ensures `binfmt_misc` support is configured on the host per the `docker buildx` documentation here: https://github.com/docker/buildx#building-multi-platform-images

r0mant commented on dronegen/container_images.go:
```diff
@@ -0,0 +1,942 @@
+	UseUniqueStagingTag          bool
+}
+
+func NewTestTrigger(triggerBranch, testMajorVersion string) *TriggerInfo {
```
Is this just for testing and need to be removed before merge?

fheinecke replied:
It is just for testing but I would recommend it stays after merge to allow for easier testing in the future. Basically this (and the block of testing const you commented on below) make it significantly easier to test changes without affecting production repos. This issue stems from not having a "staging" or "testing" drone instance to test PRs on. 

Would you prefer I remove these sections?

r0mant commented on dronegen/container_images.go:
```diff
@@ -0,0 +1,942 @@
+// * `configureForPRTestingOnly` has been set to false, and `make dronegen` has been reran afterwords
+
+const (
+	configureForPRTestingOnly bool   = false
+	testingSecretPrefix       string = "TEST_"
+	testingQuayRegistryOrg    string = "fred_heinecke"
+	testingECRRegistryOrg     string = "u8j2q1d9"
+	testingECRRegion          string = "us-east-2"
+	prBranch                  string = "fred/multiarch-teleport-container-images"
+	testingECRDomain          string = "278576220453.dkr.ecr.us-east-2.amazonaws.com"
```
I presume these need to be removed before the merge?

r0mant commented on dronegen/container_images.go:
```diff
@@ -0,0 +1,942 @@
+	}
+}
+
+func NewQuayContainerRepo(dockerUsername, dockerPassword string) *ContainerRepo {
```
Out of scope of this PR, but what are our plans for getting rid of quay.io completely? @logand22 maybe?

logand22 replied:
I wasn't sure on our plans to cut out quay.io completely. There are still quite a lot of pulls for it.

r0mant commented on dronegen/container_images.go:
```diff
@@ -0,0 +1,942 @@
+	return i.Repo == ""
+}
+
+type Product struct {
```
Can you organize the code a little better? Instead of placing everything in a single giant file, I would for example move everything related to products into, say, product.go, container repo related stuff into another file, etc. Would be easier to follow when related things are grouped together. Maybe even put them all under a subpackage e.g. dronegen/multiarch or something.

r0mant commented on dronegen/container_images.go:
```diff
@@ -0,0 +1,942 @@
+			gccPackage := ""
+			compilerName := ""
+			switch arch {
+			case "x86_64":
+				fallthrough
+			case "amd64":
```
Just do
```suggestion
			case "amd64", "x86_64":
```

r0mant commented on dronegen/container_images.go:
```diff
@@ -0,0 +1,942 @@
+			case "amd64":
+				gccPackage = "gcc-x86-64-linux-gnu"
+				compilerName = "x86_64-linux-gnu-gcc"
+			case "i686":
+				fallthrough
+			case "i386":
```
```suggestion
			case "i386", "i686":
```

fheinecke replied:
Thanks - fixed pending push

r0mant commented on dronegen/container_images.go:
```diff
@@ -0,0 +1,942 @@
+			case "i386":
+				gccPackage = "gcc-multilib-i686-linux-gnu"
+				compilerName = "i686-linux-gnu-gcc"
+			case "aarch64":
+				fallthrough
+			case "arm64":
```
```suggestion
			case "arm64", "aarch64":
```

r0mant commented on dronegen/container_images.go:
```diff
@@ -0,0 +1,942 @@
+	Product    *Product
+}
+
+type ContainerRepo struct {
```
Please add Godocs on all public types to explain what they're for. E.g .`Product`, `ContainerRepo`, `Image`, `ImageTag`, etc. It's not always clear what they represent.

This is where better code organization would also help i.e. ideally we'd have something like:

```
dronegen/
  multiarch/
    product.go
    containerrepo.go
    image.go
    cron.go
    ...
```

Probably not every single struct deserves to be in a separate file, but you get the idea.

r0mant commented on dronegen/common.go:
```diff
@@ -49,7 +49,7 @@ var (
 		Event:  triggerRef{Include: []string{"promote"}},
 		Target: triggerRef{Include: []string{"production"}},
-		Repo:   triggerRef{Include: []string{"gravitational/*"}},
+		Repo:   triggerRef{Include: []string{"gravitational/teleport"}},
```
Is this change going to affect teleport-private releases in any way?

r0mant commented on dronegen/container_images.go:
```diff
@@ -0,0 +1,942 @@
+	}
+}
+
+func NewCronTrigger(latestMajorVersions []string) *TriggerInfo {
```
We can't have a separate drone.yaml for cron but can we at leat factor out all cron related stuff in a separate .go file. I left a number of suggestion around here for better code org.

r0mant reviewed (changes requested):
Some suggestions around code organization and other things.

fheinecke commented on dronegen/common.go:
```diff
@@ -49,7 +49,7 @@ var (
 		Event:  triggerRef{Include: []string{"promote"}},
 		Target: triggerRef{Include: []string{"production"}},
-		Repo:   triggerRef{Include: []string{"gravitational/*"}},
+		Repo:   triggerRef{Include: []string{"gravitational/teleport"}},
```
Was not aware of that repo... reverted in upcoming commit

fheinecke commented on .drone.yml:
```diff
@@ -7210,8 +7216,2846 @@ steps:
+  - docker buildx build --push --builder "teleport-operator-v10-arm-builder" --platform
+    "linux/arm" --tag "drone-docker-registry:5000/teleport-operator:$DRONE_TAG-arm"
+    --file "/go/src/github.com/gravitational/teleport/operator/Dockerfile" --build-arg
+    "COMPILER_PACKAGE=gcc-arm-linux-gnueabihf" --build-arg "COMPILER_NAME=arm-linux-gnueabihf-gcc"
```
Missed that when migrating PRs - fixed

fheinecke commented on .drone.yml:
```diff
@@ -7210,8 +7216,2846 @@ steps:
+  image: docker
+  commands:
+  - apk add --no-cache aws-cli
+  - TIMESTAMP=$(date -d @"$DRONE_BUILD_CREATED" '+%Y%m%d%H%M')
```
Thanks - fixed in upcoming commit

Commit:
PR fixes

Commit:
Moved internal tools Go version to constant

Commit:
Separated container images gofile into multiple files

Commit:
Moved testing comment

Commit:
Added licenses

Commit:
Reorganized and added docs for container images

r0mant commented on dronegen/container_image_triggers.go:
```diff
@@ -0,0 +1,159 @@
+	UseUniqueStagingTag          bool
+}
+
+func NewTagTrigger(branchMajorVersion string) *TriggerInfo {
```
Nit: Can you add godocs on all public methods with some explanations on what they're for and what they do? E.g. `NewTagTrigger creates a new trigger that will execute when a new tag is pushed to Teleport repository, build multi-arch container images and push them to staging repositories`, something like that (feel free to tweak for accuracy).

r0mant commented on dronegen/container_images_testing.go:
```diff
@@ -0,0 +1,85 @@
+const (
+	configureForPRTestingOnly bool   = false
+	testingSecretPrefix       string = "TEST_"
+	testingQuayRegistryOrg    string = "fred_heinecke"
+	testingECRRegistryOrg     string = "u8j2q1d9"
+	testingECRRegion          string = "us-east-2"
+	prBranch                  string = "fred/multiarch-teleport-container-images"
```
Please change these before merge to not point to your personal org/branch.

Commit:
Merge branch 'master' into fred/multiarch-teleport-container-images

Commit:
Moved const to correct file

Commit:
Tag trigger logic test

Commit:
Testing specific fix

Commit:
Moved testing to v10.3.2

Commit:
Make semver dirs

Commit:
Refactored local registry name/socket

Commit:
Merged previous dockerfile changes

Commit:
Added TARGETOS TARGETARCH args

Commit:
Updatd tag to testing tag

Commit:
Promotion logic test

Commit:
Promotion fixes

Commit:
Testing specific fix

Commit:
Removed prerelease check for testing

Commit:
Added staging login commands to promote

Commit:
Fixed missing credentials on promotion pull

Commit:
Rerun tag test with new "full" semver

Commit:
Made staging builds only publish full semver

Commit:
Added semver logging command

Commit:
Empty commit to trigger Drone

Commit:
Promotion test

Commit:
Fixed preceeding v on promote pull

Commit:
Empty commit to trigger Drone

Commit:
Re-enabled verify not prerelease step on promote

Commit:
Cron trigger test

Commit:
Testing fix

Commit:
Testing fix 2

Commit:
Added sleep timer on docker buildx build

fheinecke commented:
Test runs prior to merging PR:
* Tag logic with empty container repo (v1.2.3-fred.3): https://drone.platform.teleport.sh/gravitational/teleport/16624
* Tag logic again with the same tag (v1.2.3-fred.4): https://drone.platform.teleport.sh/gravitational/teleport/16626
* Tag logic with the same tag, via tag delete/recreate + empty commit after (v1.2.3-fred.4): https://drone.platform.teleport.sh/gravitational/teleport/16629
* Promote logic with exit due to prerelease: https://drone.platform.teleport.sh/gravitational/teleport/16612
* Promote logic with empty container repos (v1.2.3-fred.4): https://drone.platform.teleport.sh/gravitational/teleport/16633
* Promote logic with the same tag, via empty commit (v1.2.3-fred.4): https://drone.platform.teleport.sh/gravitational/teleport/16634
* Cron logic test (v1.2.4-fred.4): https://drone.platform.teleport.sh/gravitational/teleport/16640

Everything is green and working. I need one more commit to disable my testing pipeline then I'll be merging.

Commit:
Testing cleanup

fheinecke committed:
Merge branch 'master' into fred/multiarch-teleport-container-images
